src/sage/rings/polynomial/multi_polynomial_ring_base.pyx

```diff
@@ -602,27 +599,39 @@ cdef class MPolynomialRing_base(CommutativeRing):
             sage: x + 1/u
             x + 1/u
         """
-        try:
-            P = x.parent()
-            # polynomial rings in the same variable over the any base
-            # that coerces in:
-            if isinstance(P, MPolynomialRing_base):
-                if P.variable_names() == self.variable_names():
-                    if self.has_coerce_map_from(P.base_ring()):
-                        return self(x)
-                elif self.base_ring().has_coerce_map_from(P._mpoly_base_ring(self.variable_names())):
-                    return self(x)
-
-            elif isinstance(P, polynomial_ring.PolynomialRing_generic):
-                if P.variable_name() in self.variable_names():
-                    if self.has_coerce_map_from(P.base_ring()):
-                        return self(x)
-
-        except AttributeError:
-            pass
-
-        # any ring that coerces to the base ring of this polynomial ring.
-        return self(self.base_ring().coerce(x))
+        base_ring = self.base_ring()
+        if other is base_ring:
+            # Because this parent class is a Cython class, the method
+            # UnitalAlgebras.ParentMethods.__init_extra__(), which normally
+            # registers the coercion map from the base ring, is called only
+            # when inheriting from this class in Python (cf. Issue #26958).
+            return self._coerce_map_from_base_ring()
+        f = self._coerce_map_via([base_ring], other)
+        if f is not None:
+            return f
+
+        # polynomial rings in a subset of variables
+        # over the any base that coerces in
+        if isinstance(other, MPolynomialRing_base):
+            if self is other:
+                return True
+            n = other.ngens()
+            check =  (self.ngens() >= n and
+                      self.variable_names()[:n] == other.variable_names())
```

Comment:
Isn't this assuming that the subset will be the first `n`?

Reply:
Previously the coercion system wasn't entirely self-consistent either. https://github.com/sagemath/sage/issues/39126

We can probably fix this first (keep the old non-self-consistent behavior unchanged) _then_ change it later.

Reply:
I have kept the hypothesis that the subset of variables are the initial ones, and changed the documentation accordingly. This is a progress on what was there before (only exactly the same variables) and still a reasonable assumption to make, with no possible round trips.